support-workers/src/main/scala/com/gu/stripe/StripeService.scala

```diff
@@ -15,16 +17,37 @@ class StripeService(config: StripeConfig, client: FutureHttpClient, baseUrl: Str
   val wsUrl = baseUrl
   val httpClient: FutureHttpClient = client
 
-  def createCustomer(token: String, currency: Currency): Future[Customer] =
-    postForm[Customer]("customers", Map(
-      "card" -> Seq(token)
-    ), getHeaders(currency))
+  def withCurrency(currency: Currency): WithCurrency = new WithCurrency(currency)
+  class WithCurrency(currency: Currency) {
 
-  private def getHeaders(currency: Currency) =
-    config.version.foldLeft(getAuthorizationHeader(currency)) {
-      case (map, version) => map + ("Stripe-Version" -> version)
-    }
+    def post[A](
+      endpoint: String,
+      data: Map[String, Seq[String]]
+    )(implicit decoder: Decoder[A], errorDecoder: Decoder[StripeError], ctag: ClassTag[A]): Future[A] =
+      StripeService.super.postForm[A](endpoint, data, getHeaders())
+
+    private def getHeaders() =
+      config.version.foldLeft(getAuthorizationHeader()) {
+        case (map, version) => map + ("Stripe-Version" -> version)
+      }
+
+    private def getAuthorizationHeader() =
+      Map("Authorization" -> s"Bearer ${config.forCurrency(Some(currency)).secretKey}")
+
+  }
+}
+
+object CreateCustomer {
```

Comment:
I wanted to split out the abstract definitions of the operations from the mechanism to actually make the calls - otherwise before long we end up with a lot of methods all more coupled than they should be.
However I'm open to other ways of doing it syntactically, as this specific implementation might be a bit too "FP" for the existing codebase.

Reply:
I can see what you're trying to achieve, but this feels considerably less easy to read than the previous version. If the original class had a lot of methods then I could see the value more but it only has one, so I think introducing another class and an object is overkill.

Reply:
You could leave the existing class as is pass the `StripePaymentFields` object into `createCustomer` and match on it there to work out whether to pass `source` or `payment_method` into the Stripe call

Reply:
yeah I think I'm looking at it from a purist point of view (i.e. thing X should only know about one concept) wheras you are looking for a more fewer things point of view (don't have too many "things" and fit the concepts where it makes sense at the moment and just keep the structure fluid as the balance changes)
I will go through again and tweak it back and see what it looks like.